### ISSUE TITLE
Lint request: `char_is_the_wrong_abstraction`

### ISSUE THREAD

**Issue by SoniEx2** [rust-lang/rust-clippy#6845]
### What it does

We want a restriction lint that prevents using `char` and `std::str::Chars`. We want to discourage ppl contributing to our code from using `char` because we believe it's just the wrong abstraction, and would rather it just errored so we can explain to ppl that things like grapheme clusters and whatnot exist and they should be preferred. [The std suggests you may want to look into grapheme clusters instead of using `str.chars`](https://doc.rust-lang.org/std/primitive.str.html#method.chars), which is a great suggestion but unfortunately it's just not made prominent enough, and this would fix that. Further, even the std documentation often passes the wrong idea about `char`-based APIs, leading to bugs in user code (this isn't meant to say those APIs should be deprecated, or those docs shouldn't be fixed. it's just mean to show that even std makes mistakes, so `char` is just a bad idea in general).

### Categories (optional)

- Kind: restriction
- Internationalization

### Drawbacks

Being able to use an `fn(char)->bool` as a `Pattern` is actually somewhat useful, but it's easy enough to slap an `#[allow]` on those so it's not a major drawback.

### Example

```rust
char
```

Could be written as:

```rust
SmallString<[u8; 8]>
```

### Motivation

See also https://github.com/kdeldycke/awesome-falsehood#internationalization and in particular https://eev.ee/blog/2015/09/12/dark-corners-of-unicode/ and https://manishearth.github.io/blog/2017/01/14/stop-ascribing-meaning-to-unicode-code-points/

**Comment by camsteffen** [comment]
I can see the usefulness of such a restriction lint. But it is unnecessary to assert that `char` is "just the wrong abstraction". For many use cases, it is a perfect abstraction. The lint could be named `char_for_i18n` (detects usage of `char` that is potentially insufficient internationalization and grapheme clusters could be used instead). But maybe it is reductive to say that grapheme clusters are only needed for internationalization?

**Comment by SoniEx2** [comment]
grapheme clusters also don't take e.g. terminal columns into account so they're not necessarily the correct abstraction either. so it's reductive to say grapheme clusters are sufficient for internationalization.

**Comment by camsteffen** [comment]
Okay. The lint will need to somehow be more specific and constructive than "char is wrong" to be added to Clippy.
